plugin/pkg/admission/noderestriction/admission.go

```diff
@@ -146,10 +147,10 @@ func (c *nodePlugin) Admit(a admission.Attributes) error {
 		return admission.NewForbidden(a, fmt.Errorf("disabled by feature gate %s", features.NodeLease))
 
 	case csiNodeInfoResource:
-		if c.features.Enabled(features.CSICRDAutoInstall) {
+		if c.features.Enabled(features.KubeletPluginsWatcher) {
```

Comment:
this seems more likely to be correct, but want eyes on this from the storage reviewers - is this the thing that controls whether kubelets make use of the CSINodeInfo API? using this here also means this gate will need to be set for the apiserver.

Reply:
This looks right to me too. Only when plugin registration is enabled do we attempt to write this object.

Reply:
ok. be sure to document the feature gate is required on the apiserver

Reply:
My understanding of the feature gates are:

- CSICRDAutoInstall: gates the CSINodeInfo object
- KubeletPluginsWatcher: gates the feature to have kubelet register plugins instead of the sidecar

Should we check for both feature gates in this PR?

Reply:
Yup, the logic for kubelet-based driver registration is only triggered if the overall plugin registration system is enabled

Reply:
I'd like to avoid a proliferation of flags, so no go on a third flag. If Kubelet Plugin Registration is enabled. This is the way to do registration -- makes sense to me to couple it to that flag. I also thought about checking both flags but it seems to me this particular code shouldn't really care if the CRD is installed or not.

Reply:
how do we reconcile KubeletPluginsWatcher being beta, and this being alpha?

Reply:
The way I see it is `CSICRDAutoInstall` is a guarantee that k8s will try its best to make sure the CRD is installed. It's not necessarily a gate on the CSINodeInfo object itself.

@msau42 if by reconciling you mean how we handle the fact that `KubeletPluginsWatcher` is enabled by default and `CSICRDAutoInstall` is disabled by default, the update logic `AddNodeInfo()` should handle it gracefully.

Reply:
> I think KubeletPluginsWatcher may be moving to beta

@msau42 if enabling that causes kubelets to start updating their labels, that's problematic. today, kubelets only label on initial registration. xref https://github.com/kubernetes/kubernetes/pull/68267 (which I thought could wait until early in 1.13 to go in, given the kublet is not updating its own labels currently):

> *Use cases like device plugins and CSI drivers intend to use self-added labels to express topology, which is fine, but those use cases need to be coordinated with controls the cluster-admin can use to ensure labels remain useful for partitioning a cluster along security-related boundaries (rather than starting to have kubelets set lots of labels and trying to wrap policy around that later).*

if we ship a release in which kubelets update their own labels unrestricted, our hands are tied for a year for closing the current gap to require cluster-admin opt-in for the labels they want to allow.